src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSolutionStatusServiceFactory.cs

```diff
@@ -0,0 +1,90 @@
+{
+    [ExportWorkspaceServiceFactory(typeof(ISolutionStatusService), ServiceLayer.Host), Shared]
+    internal class VisualStudioSolutionStatusServiceFactory : IWorkspaceServiceFactory
+    {
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioSolutionStatusServiceFactory(IThreadingContext threadingContext)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            if (workspaceServices.Workspace is VisualStudioWorkspace vsWorkspace)
+            {
+                var experimentationService = vsWorkspace.Services.GetService<IExperimentationService>();
+                if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.PartialLoadMode))
+                {
+                    // don't enable partial load mode for ones that are not in experiement yet
+                    return SolutionStatusService.Default;
+                }
+
+                // only VSWorkspace supports partial load mode
+                return new Service(_threadingContext, vsWorkspace);
+            }
+
+            return SolutionStatusService.Default;
+        }
+
+        private class Service : ISolutionStatusService
+        {
+            private readonly IThreadingContext _threadingContext;
+            private readonly VisualStudioWorkspace _workspace;
+
+            public Service(IThreadingContext threadingContext, VisualStudioWorkspace workspace)
+            {
+                // until we get new platform API, use legacy one that is not fully do what we want
+                _threadingContext = threadingContext;
+                _workspace = workspace;
+            }
+
+            public async System.Threading.Tasks.Task WaitForAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                if (_workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_ForceDelay))
+                {
+                    await System.Threading.Tasks.Task.Delay(_workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_DelayInMS)).ConfigureAwait(false);
+                }
+
+                var taskCompletionSource = new TaskCompletionSource<object>();
+
+                KnownUIContexts.SolutionExistsAndFullyLoadedContext.WhenActivated(() => taskCompletionSource.SetResult(null));
+
+                await taskCompletionSource.Task.ConfigureAwait(false);
+            }
+
+            public System.Threading.Tasks.Task WaitForAsync(Project project, CancellationToken cancellationToken)
+            {
+                return WaitForAsync(project.Solution, cancellationToken);
+            }
+
+            public Task<bool> IsFullyLoadedAsync(Solution solution, CancellationToken cancellationToken)
```

Comment:
why is this async?

Reply:
because real platform API will be async.